vic/drivers/cesm/cpl_mct/lnd_comp_mct.F90

```diff
@@ -0,0 +1,729 @@
+    index_x2l_Faxa_ocphidry = mct_avect_indexra(x2l, 'Faxa_ocphidry')
+    index_x2l_Faxa_ocphodry = mct_avect_indexra(x2l, 'Faxa_ocphodry')
+    index_x2l_Faxa_ocphiwet = mct_avect_indexra(x2l, 'Faxa_ocphiwet')
+    index_x2l_Faxa_dstdry1  = mct_avect_indexra(x2l, 'Faxa_dstdry1')
+    index_x2l_Faxa_dstdry2  = mct_avect_indexra(x2l, 'Faxa_dstdry2')
+    index_x2l_Faxa_dstdry3  = mct_avect_indexra(x2l, 'Faxa_dstdry3')
+    index_x2l_Faxa_dstdry4  = mct_avect_indexra(x2l, 'Faxa_dstdry4')
+    index_x2l_Faxa_dstwet1  = mct_avect_indexra(x2l, 'Faxa_dstwet1')
+    index_x2l_Faxa_dstwet2  = mct_avect_indexra(x2l, 'Faxa_dstwet2')
+    index_x2l_Faxa_dstwet3  = mct_avect_indexra(x2l, 'Faxa_dstwet3')
+    index_x2l_Faxa_dstwet4  = mct_avect_indexra(x2l, 'Faxa_dstwet4')
+
+    index_x2l_Flrr_flood    = mct_avect_indexra(x2l, 'Flrr_flood')
+
+    nflds_x2l = mct_avect_nRattr(x2l)
+
+    CALL mct_aVect_clean(x2l)
+    CALL mct_aVect_clean(l2x)
+
+  END SUBROUTINE cpl_indices_set
+
+
+  !--------------------------------------------------------------------------
+  !> @brief   Unpack the VIC domain structure
+  !--------------------------------------------------------------------------
+  INTEGER FUNCTION f_index(c_index, nx, ny)
+
+    !--- INPUT/OUTPUT PARAMETERS:
+    INTEGER(C_SIZE_T), INTENT(in)   :: c_index, nx, ny
+
+    !--- LOCAL VARIABLES:
+    INTEGER               :: row, col
+    CHARACTER(len=*), PARAMETER :: subname = '(f_index)'
+
+    !--- C to fortran index translation
+    row = INT(c_index / nx)
+    col = MOD(c_index, nx)
+    f_index = 1 + row + ny * col
+
+    RETURN
+  END FUNCTION f_index
+
+  !--------------------------------------------------------------------------
+  !> @brief   Unpack the VIC domain structure
+  !--------------------------------------------------------------------------
+  SUBROUTINE unpack_vic_domain(gindex, lon_data, lat_data, area_data, &
+                               mask_data, frac_data)
+
+    IMPLICIT NONE
+
+    !--- INPUT/OUTPUT PARAMETERS:
+    INTEGER,  DIMENSION(:), INTENT(inout) :: gindex  ! Number the local grid points
+    REAL(r8), DIMENSION(:), INTENT(inout) :: lon_data
+    REAL(r8), DIMENSION(:), INTENT(inout) :: lat_data
+    REAL(r8), DIMENSION(:), INTENT(inout) :: area_data
+    REAL(R8), DIMENSION(:), INTENT(inout) :: mask_data
+    REAL(r8), DIMENSION(:), INTENT(inout) :: frac_data
+
+    !--- LOCAL VARIABLES:
+    INTEGER :: i
+    TYPE(location_struct), DIMENSION(:), POINTER :: locations
+    CHARACTER(len=*), PARAMETER :: subname = '(unpack_vic_domain)'
+
+    !--- Associate locations pointer in local_domain structure
+    CALL c_f_pointer(local_domain%locations, locations, [local_domain%ncells])
+
+    !--- Get domain information for each cell in local domain
+    DO i = 1, local_domain%ncells
+       gindex(i) = f_index(locations(i)%global_idx, &
+                           global_domain%n_nx, global_domain%n_ny)
```

Comment:
@bartnijssen - We need to figure out how to scatter the `global_domain` structure or at least the global domain shape (`n_nx` and `n_ny`).  We need the global domain in two places at the Fortran level:
- Here, when we translate the 1d index from C order to Fortran order.
- Line 247 (above) when we calculate the full domain size / shape.

I got started on creating a MPI type for the `domain_struct` but got hung up on the `locations` member.  It seems difficult to create an MPI type of a structure with an allocatable array (pointer) inside. I understand why, I'm curious though if you have a preferred way to proceed in this area.  

The good news, after duct taping this section of code together, we're through the lnd_init step.  Not as good news is we are crashing in at the driver cesm driver level now. More work to be done.

refs:
- http://stackoverflow.com/questions/10419990/creating-an-mpi-datatype-for-a-structure-containing-pointers
- http://stackoverflow.com/questions/19637159/creating-structs-and-unions-on-mpi/19641945#19641945